Add Hikka Tracker Integration

Lorg0n opened the pull request:
Add Hikka tracker integration
Closes https://github.com/mihonapp/mihon/issues/1145

Add support for Hikka (https://hikka.io/), a Ukrainian online encyclopedia for manga and etc.

Changes implemented:
- Added new Hikka tracker with ID 10
- Implemented OAuth authentication flow with auto token refresh
- Added full sync support for:
  - Reading status (Reading, Completed, On Hold, etc.)
  - Chapter progress
  - Scores (0-10 scale)
  - Manga search and linking
- Created necessary data models and API integration
- Added UI elements for tracker settings

All changes have been tested:
- [x] Authentication flow
- [x] Manga search and linking
- [x] Status/progress syncing
- [x] Error handling

### Images
| Settings Screen | Manga Track Dialog |
| ------- | ------- |
| ![image](https://github.com/user-attachments/assets/c86d3109-7834-4428-906b-0e5497717ee9) | ![image](https://github.com/user-attachments/assets/c5dc020a-68ec-4ddf-ac66-bace0ac60b58) |

Lorg0n committed:
feat: add Hikka logo to resources

Lorg0n committed:
feat: add core Hikka tracker classes and DTOs for API requests

Lorg0n committed:
feat: update AndroidManifest for auth support, modify auth-related classes, and add Hikka to tracker list

Lorg0n committed:
feat: improve Hikka authorization

Lorg0n committed:
feat: improve Hikka authorization

Lorg0n committed:
feat: more optimization Hikka authorization

Lorg0n committed:
ref: added more specificity and removed unnecessary code in Hikka Api

Lorg0n committed:
fix: corrected function names and added reread logic

Lorg0n committed:
fix: slightly optimised the function of string to number conversion and fixed the check

Lorg0n committed:
fix: recode the rereading logic in HikkaApi

Lorg0n committed:
fix: fixed a bug where the server returned a 404 error due to receiving data in HikkaApi

Smol-Ame commented:
Include "Closes #1145" in your PR description to link to the feature request issue

AntsyLich commented:
How does the auth flow work for this site?

Lorg0n commented:
> How does the auth flow work for this site?

I didn't mean anything special, just standard OAuth authorization.

AntsyLich commented:
Ok I was ambiguous. How does one create a oauth client or get the reference?

Lorg0n commented:
> Ладно, я был двусмысленным. Как создать клиент oauth или получить ссылку?

https://hikka.io/settings/applications

Lorg0n committed:
ref: optimized imports and reduced the length of lines

AntsyLich commented:
Trying to create a new client I'm getting the error "Invalid field endpoint in request body"

AntsyLich commented:
Ok so the issue is with using the custom redirect url

Lorg0n committed:
ref: optimized imports and refactor code

AntsyLich commented:
Blocked by https://github.com/hikka-io/hikka/pull/347

Lorg0n commented:
I'll redo the authorisation to work after the changes then

Lorg0n committed:
feat: changed authorization principle, removed dependence on third-party server

Lorg0n committed:
fix: remove unused import

Lorg0n commented:
Okay, I changed the authorization. My local tests went fine
![image](https://github.com/user-attachments/assets/815dea53-8402-4f42-95db-0725b95f5df6)

Lorg0n commented:
I don't understand why the test fails, even though everything is fine in my fork. I didn't touch MangaScreenModel.kt at all

AntsyLich commented:
Yeah ignore that bit. It passes lint previously but suddenly started to flag from today.

Lorg0n committed:
fix: fixed a bug in DTO that caused a data parsing error

Lorg0n committed:
fix: fixed a bug in DTO that caused a data parsing error in manga search

Lorg0n committed:
fix: update the API link by replacing the deprecated one

Lorg0n committed:
feat: changed the buffer time to check the token for expiry

Lorg0n committed:
ref: deleted unnecessary fields in the HKUser

Lorg0n committed:
ref: deleted deleted unnecessary imports

Lorg0n committed:
feat: change logic of binding track from Hikka source

Lorg0n committed:
ref: deleted unnecessary comment

Lorg0n committed:
style: nothing change

AntsyLich commented:
Rebase please

MajorTanya reviewed (changes requested):
These are some code things I found, going to play around with the actual integration for a bit now.

MajorTanya commented:
I just saw that [Hikka merged support for start and end dates](https://github.com/hikka-io/hikka/pull/362) for manga. If you could add that once it's running in production it would be really cool. If I understand their code correctly, this is about starting and ending dates for the title itself, not the user's starting and ending dates of reading, so this would be used in the search like other trackers do.

Lorg0n commented on app/src/main/java/eu/kanade/tachiyomi/data/track/hikka/HikkaApi.kt:
```diff
@@ -0,0 +1,217 @@
+                    },
+                )
+                put("query", query)
+                put(
+                    "sort",
+                    buildJsonArray {
+                        add("score:asc")
+                        add("scored_by:asc")
+                    },
+                )
```
Unfortunately, hikka doesn't have the ability to sort by best match. In theory you can leave this field just empty, but then in the query for example for `naru` in the first item in general unknown manga with a rating of 0.

Lorg0n replied:
Oh, I didn't realise what question you were asking me. I understand, yes, I guess I forgot to change it to standard during debugging

Lorg0n commented:
Unfortunately [this update](https://github.com/hikka-io/hikka/pull/362) only applies to a request for information about a specific manga and does not give that data in a search query

Lorg0n commented on app/src/main/java/eu/kanade/tachiyomi/data/track/hikka/HikkaApi.kt:
```diff
@@ -0,0 +1,217 @@
+    companion object {
+        const val BASE_API_URL = "https://api.hikka.io"
+        const val BASE_URL = "https://hikka.io"
+        private const val SCOPE = "readlist,read:user-details"
```
Yes, "readlist" is the **read** AND **update** permission for the read manga.

Lorg0n commented on app/src/main/java/eu/kanade/tachiyomi/data/track/hikka/Hikka.kt:
```diff
@@ -0,0 +1,172 @@
+
+class Hikka(id: Long) : BaseTracker(id, "Hikka"), DeletableTracker {
+
+    companion object {
+        const val READING = 0L
+        const val COMPLETED = 1L
+        const val ON_HOLD = 2L
+        const val DROPPED = 3L
+        const val PLAN_TO_READ = 4L
+        const val REREADING = 5L
+
+        private val SCORE_LIST = IntRange(0, 10)
+            .map(Int::toString)
+            .toImmutableList()
+    }
```
Kind of not. At least companion object is at the bottom only in `Bangumi`. In the others, it's on top everywhere else (`MyAnimeList`, `Shikimori`, `MangaUpdates`, etc.).

MajorTanya replied:
Oh, lol. Guess I got too much Bangumi on my brain after my recent firestorm of Bangumi PRs. Ignore and carry on.

Lorg0n committed:
ref: simplify methods using expression bodies

Lorg0n committed:
fix: changed the sorting in the search to from higher score to lower score
ref: optimize imports and reconsider domain Track usage

Lorg0n committed:
ref: move annotations above properties and optimize string formatting in DTO

Lorg0n committed:
fix: auth issue caused by null value

Lorg0n committed:
style: add support of Hikka.io to README.md

AntsyLich commented:
Merge in the latest change from main

Lorg0n committed:
Merge branch 'main' into Add-Hikka-Tracker

MajorTanya commented on app/src/main/java/eu/kanade/presentation/more/settings/screen/SettingsTrackingScreen.kt:
```diff
@@ -168,6 +169,12 @@ object SettingsTrackingScreen : SearchableSettings {
                         logout = { dialog = LogoutDialog(trackerManager.bangumi) },
                     ),
+                    Preference.PreferenceItem.TrackerPreference(
+                        title = trackerManager.hikka.name,
```
This doesn't compile. Remove the `title` line and it should be good.

MajorTanya replied:
And in case you're worried, it still shows Hikka in the Tracking screen.

Lorg0n replied:
> And in case you're worried, it still shows Hikka in the Tracking screen.

Yes, thanks, the problem appeared because I wrote back when all trackers had a title. Fixed it now

MajorTanya reviewed (changes requested):
~Can't find anything breaking or broken, and the code appears to be nice too. Thanks for the good work.~

MajorTanya commented:
Somehow messed up with hiding the comment, it is not outdated. Please take a look at the requested change.

Lorg0n committed:
fix: resolved the Hikka title bug in SettingsTrackingScreen.kt

MajorTanya reviewed (approved):
I think this time for real I can't find anything broken. It appeared to work once the change in b34745342ab550470e53f84184d232b4fcb45009 was applied. Didn't do super in-depth edge case testing but cursory adding, moving, scoring, etc all seemed to work. 👍

Lorg0n committed:
chore: apply spotless formatting

AntsyLich commented:
The PR as is fine but I'll delay merging this until a small DB rework so we won't require workarounds like this
![image](https://github.com/user-attachments/assets/6a967631-f59e-4ed5-8181-500a73bffc8a)

Estimating a week at max

Lorg0n commented:
I'll ask the Hikka developers about a unique number(id) for the manga, I think I'll solve the problem

AntsyLich commented:
In case you misunderstood this is something I intent to fix Mihon side since more tracker we add in the future may not have numeric ID

Lorg0n commented:
Okay, I got it

AntsyLich commented:
You might want to include the new stuff from #1810

Lorg0n committed:
Merge branch 'main' into Add-Hikka-Tracker

Lorg0n committed:
feat: add start date in manga search and improve type of content

Lorg0n committed:
chore: apply Spotless formatting

Lorg0n commented:
Add `start date` and `type` of content. Unfortunately, a search of the Hikka site doesn't yield any additional data to add here.
![image](https://github.com/user-attachments/assets/8e7b0ebc-f6d6-45cd-a017-3a5d28a60761)

MajorTanya commented:
> Add `start date` and `type` of content. Unfortunately, a search of the Hikka site doesn't yield any additional data to add here. ![image](https://private-user-images.githubusercontent.com/37798530/420126493-8e7b0ebc-f6d6-45cd-a017-3a5d28a60761.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEzMDQ0NjUsIm5iZiI6MTc0MTMwNDE2NSwicGF0aCI6Ii8zNzc5ODUzMC80MjAxMjY0OTMtOGU3YjBlYmMtZjZkNi00NWNkLWEwMTctM2E1ZDI4YTYwNzYxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzA2VDIzMzYwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUxYzBjYTYyZTc4NjY0NmZmMTljNDNmNjQzOTUwZjkxOTliZmZiMjc1OTBhNmU0NzhjODhhMWQ4ZmY3ZTIzZTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.sgiKppROQrBi2ChBhtESpPjjy1yhG1oV2NNjc3W1Ors)

tragic but thanks 👍

Lorg0n commented:
> The PR as is fine but I'll delay merging this until a small DB rework so we won't require workarounds like this ![image](https://private-user-images.githubusercontent.com/59261191/417758904-6a967631-f59e-4ed5-8181-500a73bffc8a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIxNjkzODgsIm5iZiI6MTc0MjE2OTA4OCwicGF0aCI6Ii81OTI2MTE5MS80MTc3NTg5MDQtNmE5Njc2MzEtZjU5ZS00ZWQ1LTgxODEtNTAwYTczYmZmYzhhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE2VDIzNTEyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgxNTg3YzVkMzQyYTM5NzZmZDMzZWQ1YjFlNWQ5NjZmNzU2M2JlMjc5MjllMzFkNjlhOTI2OTIyOWY5MjM0MTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.haFpnYXahAFf-_soIw3JpsHrjEieAhIflK2_91Yde4Y)
> 
> Estimating a week at max

Any news on this so far?

prem-k-r commented on app/src/main/java/eu/kanade/tachiyomi/data/track/hikka/Hikka.kt:
```diff
@@ -0,0 +1,155 @@
+
+    override fun getLogoColor(): Int = Color.rgb(0, 0, 0)
+
+    override fun getLogo(): Int = R.drawable.ic_tracker_hikka
```
Rest of the trackers' logos are in the `drawable-nodpi` folder. Maybe move this one there too for consistency.

Lorg0n replied:
Yeah, it's strange. In Android Studio, the two folders appear as one folder

Lorg0n replied:
Thanks, I'm new to Android Studio

Lorg0n committed:
style: change Hikka icon location

Lorg0n committed:
fix(hikka): merge main + resolve conflict in TrackLoginActivity

Lorg0n committed:
Merge branch 'main' into Add-Hikka-Tracker

Lorg0n committed:
ref: refine hikka tracking logic and improve metadata sync

- Date Sync: Enabled 'supportsReadingDates' and implemented correct Unix timestamp conversion (API seconds <-> Android milliseconds) for both sending and receiving.
- Search Enhancement: Optimized HKManga DTO to populate user tracking progress (status, chapters, score, and dates) directly from the search results 'read' array.
- Metadata Binding: Refined 'bind' and 'refresh' methods to ensure dates and scores are fully synchronized between the server and local database.
- Data Integrity: Added nullable handling for 'startDate' and 'endDate' in HKRead DTO to prevent parsing errors when dates are missing.
- UI Update: Replaced the basic tracker icon with a high-quality Vector Drawable (brand_hikka) featuring a linear gradient and official styling.
- Logic Cleanup: Removed unnecessary 'watch' array checks from Manga DTO to strictly follow the manga API schema.
- Maintenance: Fixed potential logic inconsistencies after merging with the main branch.

Lorg0n committed:
ref(hikka): fix typo

Zeddy-Coder commented:
Hope 0.20 will finally merge this pr

AntsyLich committed:
Merge branch 'main' into Add-Hikka-Tracker

Lorg0n committed:
Merge branch 'main' into Add-Hikka-Tracker

Lorg0n committed:
Merge branch 'Add-Hikka-Tracker' of https://github.com/Lorg0n/mihon-hikka into Add-Hikka-Tracker

AntsyLich commented on app/src/main/java/eu/kanade/tachiyomi/data/track/hikka/HikkaApi.kt:
```diff
@@ -0,0 +1,219 @@
+        private const val CLIENT_REFERENCE = "49eda83d-baa6-45f8-9936-b2a41d944da4"
+        private const val CLIENT_SECRET = "8Zxzs13Pvikx6m_4rwjF7t2BxxnEb0wWtXIRQ_68HyCvmdhGE9hdfz" +
+            "SL1Pas4h927LaV2ocjVoc--S_vmorHEWWh42Z_z70j-wSFYsraQQ98" +
+            "hiOeTH2BaDf77ZcA9W5Z"
```
```suggestion
        private const val CLIENT_REFERENCE = "598ef1f5-b9d2-4e66-8b65-06949d5e14fc"
        private const val CLIENT_SECRET = "OKwzrNOZxq40psFgfcCUYddnvaeZWDnd34rt7fdcB5GmHoBBQuNTWX" +
            "61sZs8KECEWVXtMUDtq8QC4t9WX4DwWWYLXEVlgnlUXGT1fWCb-18c" +
            "Zd2m8Co-8HN6JQcjoP-B"
```

AntsyLich committed:
Update client reference and secret

AntsyLich committed:
Merge branch 'main' into Add-Hikka-Tracker

AntsyLich committed:
Update CHANGELOG.md

AntsyLich committed:
tweak icon

MajorTanya commented on app/src/main/java/eu/kanade/tachiyomi/data/track/hikka/HikkaInterceptor.kt:
```diff
@@ -0,0 +1,48 @@
+        val currAuth = oauth ?: throw Exception("Hikka: You are not authorized")
+
+        if (currAuth.isExpired()) {
+            val refreshTokenResponse = chain.proceed(HikkaApi.refreshTokenRequest(currAuth.accessToken))
+            if (!refreshTokenResponse.isSuccessful) {
+                refreshTokenResponse.close()
+                hikka.logout()
+                throw Exception("Hikka: The token is expired")
+            }
+
+            val authTokenInfoResponse = chain.proceed(HikkaApi.authTokenInfo(currAuth.accessToken))
+            if (!authTokenInfoResponse.isSuccessful) {
+                authTokenInfoResponse.close()
+            }
+
+            val authTokenInfo = json.decodeFromString<HKAuthTokenInfo>(authTokenInfoResponse.body.string())
+            setAuth(HKOAuth(currAuth.accessToken, authTokenInfo.expiration, authTokenInfo.created))
```
@Lorg0n I noticed this when I tried to debug an unclosed response issue somewhere in this, and I apologize for only bringing this up now, post-merge, but what exactly are these two requests doing? You never use the first one for anything other than checking whether it was successful. And the second one is just closed if it's unsuccessful, but afaik, the interceptor will continue its work?

Is this a "check if the refresh token is still alive" + "get a new access token" situation? If so, why is the first request content unused?

Lorg0n replied:
Hikka API has a quirk where /auth/token/info doesn't extend the token's lifetime (at least when I wrote this). To refresh it, the code has to hit any other endpoint outside /auth/, so /user/me is just the easiest one to use. The data isn't needed since the request itself is just to verify if the token is alive and automatically extend it.

And about the second request continuing after failing, that was probably just an oversight while writing and debugging. I must have missed it since it never actually crashed for me during testing

MajorTanya replied:
> Hikka API has a quirk where /auth/token/info doesn't extend the token's lifetime (at least when I wrote this). To refresh it, the code has to hit any other endpoint outside /auth/, so /user/me is just the easiest one to use.

I see, understandable then.

> And about the second request continuing after failing, that was probably just an oversight while writing and debugging. I must have missed it since it never actually crashed for me during testing

I'm actually not sure which of these requests is the one not being closed properly since it's stubbornly refusing to happen reliably or in any reproducible manner now that I've added some logcats locally...
All I've seen is an error in the search results saying that an unclosed response exists and that OKHttp couldn't do anything further until it was closed. I'm not even sure if it's the interceptor doing this or something else somewhere, though the rest of the code looks to me like it always uses the response body via `parseAs` or `use`, so those should be closed every time and this is the place with manually closed responses and only in certain conditions.

Maybe I'm just stumbling across a race condition when Hikka is a bit slow to respond or something, I don't know.

Lorg0n replied:
Maybe the leak happens because the first request (/user/me) is never closed when it's successful? It skips the if and stays open. The second one is also just missing a throw on failure

MajorTanya replied:
> Maybe the leak happens because the first request (/user/me) is never closed when it's successful? It skips the if and stays open. The second one is also just missing a throw on failure

That was my initial thought too, but I replaced the isExpired check to always be true, and it doesn't happen basically ever, even though this code path must be hit on every interaction I have with Hikka (I added dozens of titles to provoke this).

I did the always true thing because the token expires a month later but I had it happen at least three times in my build before I even started to investigate it, so it can't be about the token directly? It's all very confusing but maybe I'm missing something obvious.

Lorg0n replied:
Maybe the leak actually happens in HikkaApi.kt inside getRead()? When the app checks if a title is on the list and gets a 404 (meaning it's not there yet), it returns null, but the response is never closed. That explains why adding a bunch of new titles triggered the unclosed response errors for you

MajorTanya replied:
> Maybe the leak actually happens in HikkaApi.kt inside getRead()? When the app checks if a title is on the list and gets a 404 (meaning it's not there yet), it returns null, but the response is never closed. That explains why adding a bunch of new titles triggered the unclosed response errors for you

You misunderstood, adding all those titles _didn't_ trigger it like, ever. Whenever it did happen, it was always upon opening the Hikka search on a library entry that didn't have a Hikka entry set yet.

MajorTanya replied:
I'll open an issue about this so we can stop bothering the 8 participants in this with emails ...
